### PR TITLE
better flak with direct weapon3 tech

### PR DESCRIPTION
**Pull request by agrrr3**
this is intended to implement a shots+1 boost for flak for  MD3, Laser3, Plasma3 and DR3 techs.

for the UI I try to figure out the highest number of shots, but the necessary ShipPartMeter valueref does not have a python binding.

@o01eg could you have a look at the python binding? it is not working, I get a zero value when i trigger it with mass driver 2 tech; WEAPON_UPGRADE_SECONDARY_STAT_EFFECTS ; focs usage in ship_weapons.py

### CODE REVIEW

**Comment by o01eg** [comment]
You get zero in value of `ShipPartMeter` ?

**Comment by agrrr3** [comment]
> You get zero in value of `ShipPartMeter` ?

I am not sure. It looked like the `ShipPartMeter(part = part_name, meter = "METER_MAX_SECONDARY_STAT", ship = LocalCandidate.ID)`. After posting i checked and I think it created a ShipPartMeter valueref but it seemed it does not get evaluated. 

It created it with "SR_WEAPON_1_1", "METER_MAX_SECONDARY_STAT" and ..not sure what it used for object id...

Looking at the old parser, ValueRef::ComplexVariable<double>> should be created with (variable_name, intref ship, null,null, stringref part_name, constant string MeterToName_(meter)); i guess the parameters were good, including the meter string. (dunno about localcandidate ref).

Maybe I should filter for ships which actually do have the part. Dont know what Statistics valuerefs do if the value is not available in one of the candidates.

Where your expertise is certainly higher than mine is when trying to use a MeterType enum instead of a string.

**Comment by agrrr3** [comment]
So I solved my main problem; the string to look up for the meter should not be "METER_MAX_SECONDARY_STAT", but "MaxSecondaryStat".
This now contains everything necessary for boosting flak with MD3 tech. TODO the other *3 techs

And the one I do not know how to solve: how can I use a typed meter type instead of the string in python?

**Comment by o01eg** [comment]
> And the one I do not know how to solve: how can I use a typed meter type instead of the string in python?

I suppose that line in FOCS parser do the trick: 

```cpp
                    deconstruct_movable_(construct_movable_(new_<ValueRef::Constant<std::string>>(MeterToName_(_3))), _pass),   // string_ref2
```

Try to pass enum meter value to `ValueRef::MeterToName` function.

**Comment by agrrr3** [comment]
@o01eg seems I am missing some part of the glue;
I get 
```TypeError: No registered converter was able to produce a C++ rvalue of type MeterType from this Python object of type __MeterType```
if I
```boost::python::extract<MeterType>(kw["meter"])();```

**Comment by o01eg** [comment]
@agrrr3 You should extract to `enum_wrapper` here

**Comment by agrrr3** [comment]
hm i forgot how to invoke the ai-team

@Cjkjvfnby any suggestion if and what changes for AI are necessary?

**Comment by MichaelCourtney** [comment]
AI seems to ignore the number of shots on a flak cannon in combat rating:
https://github.com/freeorion/freeorion/blob/master/default/python/AI/CombatRatingsAI/_ship_combat_stats.py#L172-L182

Ship Designer calculates it here:
https://github.com/freeorion/freeorion/blob/master/default/python/AI/ShipDesignAI.py#L1030-L1043

As far as I'm aware it essentially ignores it, fighter damage is based on an estimation of how many fighters are destroyed rather then how much point defense the enemy has. The AI doesn't build ships with flak cannons, the designer seems to always favour arcs.

**Comment by agrrr3** [comment]
> The AI doesn't build ships with flak cannons, the designer seems to always favour arcs.

one problem which is not solved for weapons which can target multiple types (i.e. ships,planets,fighters), is that resulting damage is not split. So e.g. a single arc will look like three weapons, doing damage against ships, against planets and against fighters. Not sure how to tackle. Probably a smallish heuristic downgrade to account for part of the problem.

i will tackle double-accounting this a bit, but makes more sense in its own PR.

also the ship designer should at least consider point defense to be valuable. military decision would be the next step after those vessels get build.

> As far as I'm aware it essentially ignores it, fighter damage is based on an estimation of how many fighters are destroyed rather then how much point defense the enemy has.

currently the stats expect 20% of the fighters to survive which is OK i guess, given that currently point defense gets ignored else. This PR wont change much.

**Comment by agrrr3** [comment]
ready for merge from my side

**Comment by agrrr3** [comment]
> i will tackle double-accounting this a bit, but makes more sense in its own PR.
> 
> also the ship designer should at least consider point defense to be valuable. military decision would be the next step after those vessels get build.

added PR for those two topics: #4945

**Comment by agrrr3** [comment]
@Cjkjvfnby can we merge?

**Comment by Vezzra** [comment]
@Cjkjvfnby, poke?

**Comment by geoffthemedio** [comment]
If there's no feedback from @Cjkjvfnby or anyone else in a day or so, I'll just merge this...

File: default/scripting/techs/ship_weapons/short_range/SHP_WEAPON_2_3.focs.py
Comment (yaoyaoyao480) [inline default/scripting/techs/ship_weapons/short_range/SHP_WEAPON_2_3.focs.py:1]: 说明说明哇哈哈哈哇哈哈哈需求说明说明说明说明哇哈哈哈哇哈哈哈需求说明说明说明说明哇哈哈哈哇哈哈哈需求说明说明说明说明哇哈哈哈哇哈哈哈需求说明说明说

**Comment by Vezzra** [comment]
> If there's no feedback from @Cjkjvfnby or anyone else in a day or so, I'll just merge this...

Looks like you can go ahead...

**Comment by Blue-Award** [comment]
err something seems not quite right, @Oberlus has researched plasma 3 and got message about upgrading mass drivers, and I looked back when I put up laser 3 just for the flak upgrade, and saw that I also got message that mass drivers were updated with laser 3 tech (rather than lasers!), and I'm looking at my initial frigates and they kinda do seem to hit harder than expected

**Comment by Blue-Award** [comment]
also seems my plasmas are hitting below their their weight (got Plasma 4 but they are not hitting 108). So perhaps those upgrades laser 3 and plasma 3 indeed upgrade mass drivers instead of laser or plasma

**Comment by agrrr3** [comment]
oh no, mea culpa.
 systematic copy pasta at a blind spot.
will provide a fix for both symptons

**Pull request merged**